generators/Exercises.fs

```diff
@@ -0,0 +1,448 @@
+          Assert = this.RenderAssert canonicalDataCase }
+
+    default this.ToTestMethodBodyAssert canonicalDataCase =         
+        { Sut = this.RenderSut canonicalDataCase
+          Expected = this.RenderValueOrIdentifier (canonicalDataCase, "expected", canonicalDataCase.Expected) }
+
+    default this.ToTestMethodBodyAssertTemplate canonicalDataCase =
+        match canonicalDataCase.Expected with
+        | :? JArray as jArray when jArray.Count = 0 && not (List.contains "expected" (this.PropertiesWithIdentifier canonicalDataCase)) -> "AssertEmpty"
+        | _ -> "AssertEqual"
+
+    // Rendering of canonical data
+
+    default this.Render canonicalData =
+        canonicalData
+        |> this.ToTestClass
+        |> renderPartialTemplate "TestClass"
+
+    default this.RenderTestMethod index canonicalDataCase = 
+        canonicalDataCase
+        |> this.ToTestMethod index
+        |> renderPartialTemplate "TestMethod"
+
+    default this.RenderTestMethodBody canonicalDataCase = 
+        canonicalDataCase
+        |> this.ToTestMethodBody
+        |> renderPartialTemplate "TestMethodBody"
+
+    default this.RenderTestMethodName canonicalDataCase = 
+        match this.UseFullMethodName canonicalDataCase with
+        | false ->
+            String.upperCaseFirst canonicalDataCase.Description
+        | true -> 
+            canonicalDataCase.DescriptionPath
+            |> String.concat " - "
+            |> String.upperCaseFirst    
+
+    // Generic value/identifier rendering methods
+
+    default this.RenderValue (canonicalDataCase, key, value) = formatValue value
+
+    default this.RenderValueOrIdentifier (canonicalDataCase, key, value) =
+        let properties = this.PropertiesWithIdentifier canonicalDataCase
+
+        match List.contains key properties with
+        | true  -> this.RenderIdentifier (canonicalDataCase, key, value)
+        | false -> this.RenderValueWithoutIdentifier (canonicalDataCase, key, value)
+
+    default this.RenderValueWithoutIdentifier (canonicalDataCase, key, value) = 
+        match key with
+        | "expected" -> this.RenderExpected (canonicalDataCase, key, value)
+        | _  -> this.RenderInput (canonicalDataCase, key, value)
+
+    default this.RenderValueWithIdentifier (canonicalDataCase, key, value) = 
+        let identifier = this.RenderIdentifierWithTypeAnnotation (canonicalDataCase, key, value)
+        let value = this.RenderValueWithoutIdentifier (canonicalDataCase, key, value)
+        sprintf "let %s = %s" identifier value  
+
+    default this.RenderIdentifier (canonicalDataCase, key, value) = String.camelize key
+
+    default this.RenderIdentifierWithTypeAnnotation (canonicalDataCase, key, value) = 
+        let identifier = this.RenderIdentifier (canonicalDataCase, key, value)
+    
+        match this.IdentifierTypeAnnotation (canonicalDataCase, key, value) with
+        | Some ty -> identifier |> addTypeAnnotation ty
```

Comment:
Why `ty`?

Reply:
Because `type` is a reserverd keyword. But I'll rename, you're right.

Reply:
Ah ok that makes sense :)